### PR TITLE
ENH: special: ensure tests are thread-safe

### PR DESCRIPTION
**Pull request by andfoy**
#### Reference issue

See https://github.com/scipy/scipy/issues/20669

#### What does this implement/fix?

This is a continuation of the work done in https://github.com/scipy/scipy/pull/21496, this time, each module gets its own PR detailing changes that ensure that the tests pass under a concurrent scenario. 

This PR was split from the original diff on https://github.com/scipy/scipy/pull/21956, as per https://github.com/scipy/scipy/pull/21956#issuecomment-2505890990

#### Additional information

See the description of https://github.com/scipy/scipy/pull/21496 for more detailed information.

cc @rgommers

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
On MacOS ARM with Python `3.13.0t`, `PYTHON_GIL=0`, and incantation `python dev.py test -j 8 -t scipy/special/tests -- --parallel-threads=10` I see an improvement from 173 test failures on latest SciPy `main` to 29 failures on this branch.

The diff seems sensible enough to me, though I'm not a special function domain expert. CI failures are either seen elsewhere or minor linting complaints for other lines in touched files.

**Review by rgommers — COMMENTED** [review]
Thanks @andfoy. Largely good to go, a couple of questions.

File: scipy/special/tests/test_basic.py [hunk 1]
```diff
@@ -4280,9 +4284,16 @@ def test_legacy():
...
+@pytest.fixture
+def errstate_lock():
+    import threading
+    return threading.Lock()
+
+
 @with_special_errors
-def test_error_raising():
-    assert_raises(special.SpecialFunctionError, special.iv, 1, 1e99j)
+def test_error_raising(errstate_lock):
```
[the remark this answers is not in the document]
Comment (andfoy) [reply scipy/special/tests/test_basic.py:4294]: Indeed, the lock is related to the `sf_error` issue.

[the remark this answers is not in the document]
Comment (rgommers) [reply scipy/special/tests/test_basic.py:4293]: Thanks for confirming. I pushed a code comment that says this.

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @andfoy.  

There are still issues with error handling that result in lots of test failures, but those should all be resolved in gh-21956.

**Pull request merged**